contributors/design-proposals/network/service-topology-key.md

```diff
@@ -0,0 +1,207 @@
+
+`ExternalTrafficPolicy` was added in v1.4, but only for NodePort and external LB traffic. NodeName was added to `EndpointAddress` to allow kube-proxy to filter local endpoints for various future purposes.
+
+Based on our experience of advanced routing setup and recent demo of enabling this in Kubernetes service, this document would like to introduce a more generic way to enable "local-only service".
+
+## Proposal
+
+This proposal builds off of earlier requests to [use local pods only for kube-proxy loadbalancing](https://github.com/kubernetes/kubernetes/issues/7433) and [node-local service proposal](https://github.com/kubernetes/kubernetes/pull/28637). But, this document proposes that not only the particular "node-local" user case should be taken care, but also a more generic way should be figured out.
+
+Locality is an "user-defined" thing. When we set topology key "hostname" for service, we expect node carries different node labels on the key "hostname".
+
+Users can control the level of topology. For example, if someone run logging agent as a daemonset, he can set the "hard" topology requirement for same-host. If "hard" is not met, then just return "service not available". 
+
+And if someone set a "soft" topology requirement for same-host, say he preferred "same-host" but it's not required, he can accept other hosts when for some reasons local service's backend is not available on some host.
+
+If multiple endpoints satisfy the "hard" topology requirement, we will evaluate "how much preferred" each endpoint is by the "soft" topology requirement, and implement some algorithm like "weighted round robin".
+
+Routing decision is expected to be implemented in L3/4 VIP level such as kube proxy.
+
+## Implementation details
+
+### API changes
+
+#### Service API changes
+
+As with scheduling topology, there are currently two types of topology, called `Required` and `Preferred` which denote “hard” vs. “soft” requirements. 
+
+Similarly, we introduce `Required` and `Preferred` in service topology.
+
+An example of `required` topology would be “only visit service backends in the same zone”.
+
+An example of `preferred` topology would be “prefer to visit service backends in the same host when multiple endpoints are in the same zone”. Then "preferred" endpoints may have more routing probabilities.
+
+```go
+type Service struct {
+  Spec:
+    // If the topology requirements specified by this field are not met, the LB, such as     // kube-proxy will not pick endpoints for the service.
+    // ordering doesn't matter, it's an "AND" semantic in Required array.
+    // +optional
+    Required []TopologyTerm
```

Comment:
Having the word "Required" (and  "Preferred" in the following) is not enough self-explanatory, I would suggest something like "RequiredServiceTopology" instead:
```go
type ServiceSpec struct {
	RequiredServiceTopology []TopologyTerm
	PreferredServiceTopology []WeightedTopologyTerm
}
```
Or maybe
```go
type ServiceSpec struct {
	Topology ServiceTopology
}

type ServiceTopology struct {
	Required []TopologyTerm
	Preferred []WeightedTopologyTerm
}
```
The latter one maybe too much nested, though that's how the affinity staff defined in podSpec today.

Reply:
And to have a comparison with the flattened fields style which @thockin suggested:
```go
    Topology ServiceTopology
    // ...
  }
  type ServiceTopology struct {
      Mode ServicetopologyMode
      Key string
  }
```
the major differences are:
1. In nested (pod affinity) style, service could have both required topology routing requirements and preferred ones, e.g. user wants to restrict the request in the same AZ with client, and prefer to route traffics to same rack first, but OK to other racks if none match.
2. With the flattened fields style, a service is configured with either required or preferred routing requirements. (To represent "ignored", simply leave service.spec.topolgy nil/empty)